CloudMonitor: Fix to hide queries when used in conjunction with Expressions

aangelisc opened the pull request:
**What this PR does / why we need it**:

Fix to correctly hide Cloud Monitor queries when used with expressions. 

**Which issue(s) this PR fixes**:

Fixes #50449 

**Special notes for your reviewer**:

aangelisc committed:
Correctly pass all original query parameters on interpolation

andresmgot reviewed (approved):
Can you add a test? It's difficult to see what was the property that actually did the trick here. Thanks!

aangelisc committed:
Add datasource tests

- Update props on mock query
- Add mock instance settings
- Add util function to generate template variables
- Add datasource test file
- Correct setting of templateSrv
- Update betterer results

joshhunt commented on .betterer.results:
```diff
@@ -6266,6 +6266,12 @@ exports[`better eslint`] = {
       [0, 0, 0, "Do not use any type assertions.", "1"],
       [0, 0, 0, "Do not use any type assertions.", "2"]
     ],
+    "public/app/plugins/datasource/cloud-monitoring/__mocks__/instanceSettings.ts:5381": [
+      [0, 0, 0, "Do not use any type assertions.", "0"]
+    ],
+    "public/app/plugins/datasource/cloud-monitoring/__mocks__/utils.ts:5381": [
+      [0, 0, 0, "Do not use any type assertions.", "0"]
+    ],
```
Please rebase after this PR is merged, and re-betterer to get rid of these https://github.com/grafana/grafana/pull/53593

aangelisc committed:
Simplify test and pass templateSrv appropriately

aangelisc committed:
Merge branch 'main' into 50449-cloudmonitor-query-hiding

# Conflicts:
#	.betterer.results

aangelisc committed:
Fix lint issue

andresmgot commented on public/app/plugins/datasource/cloud-monitoring/datasource.test.ts:
```diff
@@ -0,0 +1,42 @@
+    });
+
+    it('should correctly apply template variables', () => {
+      const templateVariables = createTemplateVariables(['metricQuery.projectName']);
```
I believe this util function is hurting more than helping, it's a bit difficult to know what's happening. You are passing a path as a string, which should match a real object path to work and generates something a bit obscure (you are losing all the benefits of typings). It also relies a lot in the implementation of templateSrv, which we don't need to care.

I believe you can simplify all this test with something like:

```
    it('should correctly apply template variables', () => {
      templateSrv.replace = jest.fn().mockReturnValue('myvar');
      const query = createMockQuery({ hide: false });
      const templatedQuery = ctx.ds.interpolateVariablesInQueries([query], {});
      expect(templatedQuery[0].metricQuery.projectName).toEqual('myvar');
    });
```

and test the same thing. You can even check if `.replace` was called with the right arguments if you want to.

Also, I believe this test is not testing anything new? This test should pass with teh current main so you are not testing the bug fix?

aangelisc replied:
This test wouldn't pass on current main because of the `hide: false`. The main change in this PR is that we now spread the whole query when applying template variables to ensure none of the original query properties are lost (which is why the bug appeared).

Your implementation for the test does make more sense, I'll make some changes on this 😊

andresmgot replied:
aah, gotcha, can you add that to the test description or something? (also a test with `hide: true` to see the difference?)

aangelisc committed:
Simplify test and add type that supports parital nested objects

aangelisc committed:
Update test and remove unneeded util function

aangelisc committed:
Merge branch 'main' into 50449-cloudmonitor-query-hiding

andresmgot commented on public/app/plugins/datasource/cloud-monitoring/__mocks__/cloudMonitoringQuery.ts:
```diff
@@ -28,12 +35,17 @@ export const createMockSLOQuery: (overrides?: Partial<SLOQuery>) => SLOQuery = (
 };
 
-export const createMockQuery: (overrides?: Partial<CloudMonitoringQuery>) => CloudMonitoringQuery = (overrides) => {
+export const createMockQuery: (overrides?: Subset<CloudMonitoringQuery>) => CloudMonitoringQuery = (overrides) => {
```
curious: what's the difference between Partial and Subset? Is it to maintain the original type?

aangelisc replied:
`Partial` doesn't work for nested objects (so in this case if we pass this override to `createMockQuery`: `createMockQuery({ metricQuery: { projectName: '$testVar' } })` then we will get a type error as `Partial` allows `metricQuery` to be undefined but this does not extend to the properties within `metricQuery`. `Subset` extends `Partial` to work for nested objects in this case 😊

aangelisc committed:
Rename to avoid duplicate mock

andresmgot reviewed (approved):
great :+1:
